text/projects/Krzysztof_Pióro/Maximal_Suffix.tex

```diff
@@ -0,0 +1,138 @@
+\author{Krzysztof Pióro}
+\date{Maj 2022}
+
+\begin{document}
+
+\maketitle
+
+\section{Wstęp}
+
+Autorzy pracy pokazują prosty w opisie algorytm wyznaczający maksymalny sufiks słowa działający w czasie liniowym i stałej dodatkowej pamięci.
+
+\section{Definicje}
+
+Będziemy rozważać słowo $w$ długości $n$ indeksowane od $1$ do $n$. Dodatkowo definiujemy:
+
+\begin{itemize}
+    \item MaxSuf$(w)$ -- maksymalny sufiks słowa $w$
+    \item period$(w)$ -- najkrótszy okres słowa $w$
+    \item słowo nazywamy \textit{samo-maksymalnym} $\iff$ MaxSuf$(w) = w$.
+    \item MaxSufPos$(w)$ -- pozycja od której zaczyna się maksymalny sufiks słowa $w$
+    \item $w_1 < w_2$ -- $w_1$ jest mniejsze leksykograficznie od $w_2$
+\end{itemize}
+
+
+
+\section{Algorytm}
+
+\begin{algorithm}[H]
+    \caption{\textbf{Compute-MaxSufPos$(w)$}}
+    \begin{algorithmic}
+        \State $i := 1$; $j := 2$;
+        \While {$j \leq n$}
+            \State $k := 0$
+            \While {$j + k < n$ \textbf{and} $w[i+k] = w[j+k]$}
+                \State $k := k + 1$;
+            \EndWhile
+            \If{$w[i+k] < w[j+k]$}
+                \State $i := i + k + 1$;
+            \Else
+                \State $j := j + k + 1$;
+            \EndIf
+            \If {$i = j}$
+                \State $j := j + 1$;
+            \EndIf
+        \EndWhile
+        \State \Return $i$
+    \end{algorithmic}
+\end{algorithm}
+
+
+Algorytm oczywiście działa w stałej (dodatkowej) pamięci i w liniowej liczbie operacji.
+
+\section{Dowód poprawności}
+
+Niech $(i, j) \rightarrow (i', j')$ oznacza, że z konfiguracji $(i, j)$  w jednej iteracji przechodzimy do $(i', j')$ i niech $\rightarrow^{*}$ będzie
+domknięciem przechodnim relacji $\rightarrow$. 
+
+Pokażemy, że po każdej głównej iteracji będą zachodziły poniższe \textbf{niezmienniki} (gdzie oznaczamy $u := w[i\ldots j-1]$):
+\begin{enumerate}
+    \item $(i < j < n) \Rightarrow u$ jest \textit{samo-maksymalne} oraz $period(u) = u$
+    \item maksymalny sufiks słowa $w$ nie zaczyna się przed $i$
+\end{enumerate}
+
+
+
+\begin{proof}[\textbf{Dowód niezmienników}]
+Początkowo $i = 1, j = 2$ i niezmiennik jest spełniony
+
+Rozważmy iterację, w której $i$ jest zmienione po raz pierwszy. Pokażemy, że niezmienniki są spełnione przed tą iteracją 
+(autorzy pracy piszą tutaj tylko, że jest to "easy to see", ale my to uzasadnimy). Oczywiście drugi niezmiennik jest spełniony, zatem skupimy się na pierwszym.
+
+Rozważamy zatem przejście $(i=1, j) \rightarrow (i=1, j' = j + k + 1)$, gdzie niezmienniki zachodzą dla słowa $u = w[i \ldots j-1]$.
+
+Wtedy niech $u' := w[i\ldots j + k] = u^tva$, gdzie $v$ jest ścisłym prefiksem $u$ (być może pustym), a $a$ jest literką taką, że $va < u$.
+
+Z \textit{samo-maksymalności} słowa $u$ oraz faktu, że $va < u$ możemy wywnioskować, że słowo $u'$ również jest \textit{samo-maksymalne}.
+Podobnie możemy też wywnioskować, że słowo $u^tv$ jest \textit{samo-maksymalne} (skorzystamy z tego później). 
```

Comment:
Both can be proved within few minutes, but frankly I see them as not obvious at all.